packages/tasit-action/src/contract/Contract.test.js

```diff
@@ -334,45 +333,30 @@ describe("TasitAction.Contract", () => {
       expect(errorListener.called).to.be.false;
     });
 
-    // Non-deterministic test case
-    it.skip("should call error listener after timeout", async () => {
+    it("action should call error listener after timeout", done => {
       action = sampleContract.setValue("hello world");
       action.setEventsTimeout(100);
 
       const errorListener = sinon.fake(error => {
-        expect(error.eventName).to.equal("confirmation");
-        expect(error.message).to.equal("Event confirmation reached timeout.");
+        const { eventName, message } = error;
+        expect(eventName).to.equal("confirmation");
+        expect(message).to.equal("Event confirmation reached timeout.");
+        expect(action.subscribedEventNames()).to.deep.equal([
+          "error",
+          "confirmation",
+        ]);
         action.off("error");
-      });
-
-      const confirmationListener = sinon.fake(() => {
         action.off("confirmation");
+        done();
       });
 
-      action.on("confirmation", confirmationListener);
+      const confirmationListener = sinon.fake(() => {});
 
+      action.on("confirmation", confirmationListener);
       action.on("error", errorListener);
 
-      await action.send();
-      await action.waitForOneConfirmation();
-
-      await mineBlocks(provider, 2);
-
-      // TODO: Use fake timer when Sinon/Lolex supports it.
-      // See more:
-      //  https://github.com/sinonjs/sinon/issues/1739
-      //  https://github.com/sinonjs/lolex/issues/114
-      //  https://stackoverflow.com/a/50785284
-      await wait(action.getEventsTimeout() * 5);
-
-      await mineBlocks(provider, 2);
-
-      expect(errorListener.callCount).to.equal(1);
-      expect(confirmationListener.callCount).to.equal(1);
-      expect(action.subscribedEventNames()).to.deep.equal([
-        "error",
-        "confirmation",
-      ]);
+      // Note: No time is necessary because no additional block will be mined
```

Comment:
But if it’s calling it after a timeout, I don’t totally understand what this means here. Do you mean no block mining is necessary, as long as the event timeout is shorter than the test timeout?

Reply:
- Test case timeout (mocha) is 8 sec;
- Test case will wait until `done()` is called;
- Timeout error will be thrown if the time between blocks be > timeout;
That said, after first confirmation, since now the additional block is being called, the timeout error is being thrown.
Make sense?

Reply:
> Timeout error is thrown in practice mostly if there’s a new block without a new confirmation for that tx

Timeout error is thrown if the time between our `ethers.js` wrapper listener triggers is greater than the timeout. (I'll put the code below). It will occur for block event ('confirmation') and contract event.

> But the lack of a block also works.

Yes, because we are using [`ethers.js` block event](https://docs.ethers.io/ethers.js/html/api-providers.html#event-types) for our `confirmation` event and without new block, the timeout will be reached.

> So the timeout time period happens after the block that is mined as a result of this send, right? (This is my understanding right now.)
Or did the time period happen already and end when this send happens?

Could clarify what you've meant with "as result of this send"?


Code from `Subcription` class

```javascript
const listener = async (...args) => {
  const eventListener = this.#eventListeners.get(eventName);
  const { isRunning } = eventListener;

  if (isRunning) {
    console.info(`Listener is already running`);
    return;
  }

  this.#decorateEventListener(eventName, {
    isRunning: true,
    lastEmissionTime: Date.now(),
  });

  this._clearEventTimerIfExists(eventName);

  const timer = setTimeout(() => {
    const eventListener = this.#eventListeners.get(eventName);
    const { lastEmissionTime } = eventListener;
    const currentTime = Date.now();
    const timedOut =
      currentTime - lastEmissionTime >= this.getEventsTimeout();

    if (timedOut) {
      this._emitErrorEventFromEventListener(
        new Error(`Event ${eventName} reached timeout.`),
        eventName
      );
    }
  }, this.getEventsTimeout());

  this._setEventTimer(eventName, timer);

  await baseListener(...args);
  this.#decorateEventListener(eventName, { isRunning: false });
};

const eventListener = { listener, isRunning: false };

this.#eventListeners.set(eventName, eventListener);

this.#ethersEventEmitter.on(this._toEthersEventName(eventName), listener);
```

Reply:
> It will occur for block event ('confirmation') and contract event.

But there's only a timeout for a contract event if we set one explicitly, right?

Reply:
>> So the timeout time period happens after the block that is mined as a result of this send, right? (This is my understanding right now.)
Or did the time period happen already and end when this send happens?

> Could clarify what you've meant with "as result of this send"?

I mean that ethers.js is set to instantly mine a new block whenever a tx is broadcast. So the fact that a new block is mined is due to the `.send()`

Reply:
> > It will occur for block event ('confirmation') and contract event.
> 
> But there's only a timeout for a contract event if we set one explicitly, right?

No, not really.

Reply:
> > > So the timeout time period happens after the block that is mined as a result of this send, right? (This is my understanding right now.)
> > > Or did the time period happen already and end when this send happens?
> 
> > Could clarify what you've meant with "as result of this send"?
> 
> I mean that ethers.js is set to instantly mine a new block whenever a tx is broadcast. So the fact that a new block is mined is due to the `.send()`

> > > So the timeout time period happens after the block that is mined as a result of this send, right? (This is my understanding right now.)

Right.

Reply:
> > > It will occur for block event ('confirmation') and contract event.
> > 
> > 
> > But there's only a timeout for a contract event if we set one explicitly, right?
> 
> No, not really.

In that case, as of right now we can't listen for `ExecutionFailed` forever for a contract. Is that correct?

Reply:
> > > > So the timeout time period happens after the block that is mined as a result of this send, right? (This is my understanding right now.)
> > > > Or did the time period happen already and end when this send happens?
> > 
> > 
> > > Could clarify what you've meant with "as result of this send"?
> > 
> > 
> > I mean that ethers.js is set to instantly mine a new block whenever a tx is broadcast. So the fact that a new block is mined is due to the `.send()`
> 
> > > > So the timeout time period happens after the block that is mined as a result of this send, right? (This is my understanding right now.)
> 
> Right.

In that case, I think we can change the comment

Reply:
```suggestion
      // Note: Sending this triggers a block being mined,
      // and since no additional blocks will be mined,
      // the timeout will be reached before another confirmation occurs
```

Reply:
> In that case, as of right now we can't listen for ExecutionFailed forever for a contract. Is that correct?

Yes, that's correct. Should I create an issue to having that listening forever as default?

Reply:
Yeah, I think so, if you agree

Reply:
Comment was changed on https://github.com/tasitlabs/TasitSDK/pull/392

Reply:
> > In that case, as of right now we can't listen for ExecutionFailed forever for a contract. Is that correct?
> 
> Yes, that's correct. Should I create an issue to having that listening forever as default?

Extracted to: https://github.com/tasitlabs/TasitSDK/issues/393

Reply:
> Comment was changed on #392

Awesome, thanks

Reply:
> > Yeah, I think so, if you agree
> 
> Extracted to: #393

Perfect, resolving this